# Characterization of altered molecular mechanisms in Parkinson’s disease through cell type-resolved multi-omics analyses

Parkinson’s disease (PD) is a progressive neurodegenerative disorder. However, cell type-dependent transcriptional regulatory programs responsible for PD pathogenesis remain elusive. Here, we establish transcriptomic and epigenomic landscapes of the substantia nigra (SN) by profiling 113,207 nuclei obtained from healthy controls and PD patients. Our multi-omic data integration provides cell type annotation of 128,724 cis-regulatory elements (cREs), and uncovers cell type-specific cRE dysregulations with a strong transcriptional influence on genes implicated in PD. The establishment of high-resolution three-dimensional chromatin contact maps identifies 656 target genes of dysregulated cREs and genetic risk loci, uncovering both novel candidates and known PD risk genes. Notably, these candidate genes exhibit modular gene expression patterns with unique molecular signatures in distinct cell types, highlighting altered molecular mechanisms in dopaminergic neurons and glial cells including oligodendrocytes and microglia. Together, our single-cell transcriptome and epigenome uncover cell type-specific disruption in transcriptional regulations related to Parkinson’s diseases.
